main/src/mill/MillMain.scala

```diff
@@ -265,31 +265,19 @@ object MillMain {
                 stderr.println("Trying to load BSP server...")
                 val bspServerFuture = Future {
                   try {
-                    val bspClass = MillMain.this.getClass.getClassLoader.loadClass("mill.bsp.BSP")
-                    val method = bspClass.getMethod(
-                      "startBspServer",
-                      Seq[Class[_]](
-                        classOf[Option[Evaluator]],
-                        classOf[PrintStream],
-                        classOf[PrintStream],
-                        classOf[InputStream],
-                        classOf[os.Path],
-                        classOf[Boolean],
-                        classOf[Option[Promise[BspServerHandle]]]
-                      ): _*
-                    )
-
-                    method.invoke(
-                      null,
-                      Seq[Object](
-                        None,
-                        MillMain.initialSystemStreams.out,
-                        System.err,
-                        MillMain.initialSystemStreams.in,
-                        os.pwd / ".bsp",
-                        java.lang.Boolean.TRUE,
-                        Some(bspServerHandle)
-                      ): _*
+                    // We cannot link this class directly, as it would give us a circular dependency
+                    val bspClass = MillMain.this.getClass.getClassLoader.loadClass("mill.bsp.BspServerStarterImpl")
+                    val method = bspClass.getMethod("get")
+                    val serverStarter = method.invoke(null).asInstanceOf[BspServerStarter]
```

Comment:
Move reflection logic into `BspServerStarter` companion, paired with more info why it is done that way.